packages/core/src/lib/base_protocol.ts

```diff
@@ -28,13 +29,26 @@ export class BaseProtocol {
   }
 
   protected async getPeer(peerId?: PeerId): Promise<Peer> {
-    const { peer } = await selectPeerForProtocol(
-      this.peerStore,
-      [this.multicodec],
-      peerId
-    );
+    let peer = await this._getPeer(peerId);
+    if (!peer) {
+      await waitForRemotePeer(this.multicodec, this.peerStore);
```

Comment:
What is the likelihood of a connection finishing when line 33 is being executed? Meaning that we may miss the event for it.

Wouild probably need to  listen first and resolve if we have a peer available or if not we have an event coming.